src/vs/workbench/contrib/scm/browser/dirtydiffDecorator.ts

```diff
@@ -244,28 +264,96 @@ class DirtyDiffWidget extends PeekViewWidget {
 
+		let detail: string;
+		let display: 'none' | 'inherit';
+		if (!this.shouldUseDropdown()) {
+			detail = this.model.changes.length > 1
+				? nls.localize('changes', "{0} - {1} of {2} changes", label, providerIndex + 1, providerChanges.length)
+				: nls.localize('change', "{0} - {1} of {2} change", label, providerIndex + 1, providerChanges.length);
+			display = 'none';
+		} else {
+			detail = this.model.changes.length > 1
+				? nls.localize('multiChanges', "{0} of {1} changes", providerIndex + 1, providerChanges.length)
+				: nls.localize('multiChange', "{0} of {1} change", providerIndex + 1, providerChanges.length);
+			display = 'inherit';
+		}
+		if (this.dropdownContainer?.style) {
+			this.dropdownContainer.style.display = display;
+		}
 		this.setTitle(this.title, detail);
 	}
 
+	private switchQuickDiff(event: unknown) {
+		const newProvider = (event as { provider: string }).provider;
+		if (newProvider === this.model.changes[this._index].label) {
+			return;
+		}
+		let closestGreaterIndex = this._index < this.model.changes.length - 1 ? this._index + 1 : 0;
+		for (let i = closestGreaterIndex; i !== this._index; i < this.model.changes.length - 1 ? i++ : i = 0) {
+			if (this.model.changes[i].label === newProvider) {
+				closestGreaterIndex = i;
+				break;
+			}
+		}
+		let closestLesserIndex = this._index > 0 ? this._index - 1 : this.model.changes.length - 1;
+		for (let i = closestLesserIndex; i !== this._index; i >= 0 ? i-- : i = this.model.changes.length - 1) {
+			if (this.model.changes[i].label === newProvider) {
+				closestLesserIndex = i;
+				break;
+			}
+		}
+		const closestIndex = Math.abs(this.model.changes[closestGreaterIndex].change.modifiedEndLineNumber - this.model.changes[this._index].change.modifiedEndLineNumber)
+			< Math.abs(this.model.changes[closestLesserIndex].change.modifiedEndLineNumber - this.model.changes[this._index].change.modifiedEndLineNumber)
+			? closestGreaterIndex : closestLesserIndex;
+		this.showChange(closestIndex, false);
+	}
+
+	private shouldUseDropdown(): boolean {
+		let providersWithChangesCount = 0;
+		if (this.model.mapChanges.size > 1) {
+			const keys = Array.from(this.model.mapChanges.keys());
+			for (let i = 0; (i < keys.length) && (providersWithChangesCount <= 1); i++) {
+				if (this.model.mapChanges.get(keys[i])!.length > 0) {
+					providersWithChangesCount++;
+				}
+			}
+		}
+		return providersWithChangesCount >= 2;
+	}
+
 	protected override _fillHead(container: HTMLElement): void {
 		super._fillHead(container, true);
 
+		if (this._titleElement) {
+			this.dropdownContainer = dom.prepend(this._titleElement, dom.$('.dropdown'));
+			this.dropdown = this.instantiationService.createInstance(SwitchQuickDiffViewItem, new SwitchQuickDiffAction((event: unknown) => this.switchQuickDiff(event)),
+				this.model.quickDiffs.map(quickDiffer => quickDiffer.label), this.model.changes[this._index].label);
+			this.dropdown.render(this.dropdownContainer);
+		}
+
```

Comment:
Even though `_titleElement` is `?`, there's not reason for this code to not assume it is set. If unset, it's better that we know about it via a stack trace than possibly not know about it via the dropdown not showing up.

My suggestion here is to change the reference to `this._titleElement!`.